x-pack/plugins/apm/server/lib/traces/get_trace.ts

```diff
@@ -10,37 +10,109 @@ import { Span } from '../../../typings/es_schemas/Span';
 import { Transaction } from '../../../typings/es_schemas/Transaction';
 import { Setup } from '../helpers/setup_request';
 
-export type TraceAPIResponse = Array<Transaction | Span>;
+export interface TraceAPIResponse {
+  trace: Array<Transaction | Span>;
+  errorsPerTransaction: {
+    [transactionId: string]: number;
+  };
+}
+
+interface TraceErrorsAggBucket {
+  key: string;
+  errors: {
+    doc_count: number;
+  };
+}
+
+interface TraceErrorsAggResponse {
+  transactions: {
+    buckets: TraceErrorsAggBucket[];
+  };
+}
 
 export async function getTrace(
   traceId: string,
   setup: Setup
 ): Promise<TraceAPIResponse> {
   const { start, end, client, config } = setup;
 
-  const params: SearchParams = {
-    index: config.get('apm_oss.transactionIndices'),
+  const query = {
+    bool: {
+      filter: [
+        { term: { [TRACE_ID]: traceId } },
+        {
+          range: {
+            '@timestamp': {
+              gte: start,
+              lte: end,
+              format: 'epoch_millis'
+            }
+          }
+        }
+      ]
+    }
+  };
+
+  const traceParams: SearchParams = {
+    index: [
+      config.get('apm_oss.spanIndices'),
+      config.get('apm_oss.transactionIndices')
+    ],
     body: {
       size: 1000,
-      query: {
-        bool: {
-          filter: [
-            { term: { [TRACE_ID]: traceId } },
-            {
-              range: {
-                '@timestamp': {
-                  gte: start,
-                  lte: end,
-                  format: 'epoch_millis'
+      query
+    }
+  };
+
+  const errorsAggParams: SearchParams = {
+    index: [config.get('apm_oss.errorIndices')],
+    body: {
+      size: 0,
+      query,
+      aggs: {
+        transactions: {
+          terms: {
+            field: 'transaction.id'
+          },
+          aggs: {
+            errors: {
+              filter: {
+                term: {
+                  'processor.event': 'error'
                 }
               }
             }
-          ]
+          }
         }
       }
     }
   };
 
-  const resp = await client<Span | Transaction>('search', params);
-  return resp.hits.hits.map(hit => hit._source);
+  return Promise.all([
+    client<Span | Transaction>('search', traceParams), // query for trace items (spans & transaction)
+    client<Span | Transaction, TraceErrorsAggResponse>( // agg for transaction error count
+      'search',
+      errorsAggParams
+    )
+  ]).then(([traceResponse, errorsAggResponse]) => {
+    const trace = traceResponse.hits.hits.map(hit => hit._source);
+    const errorsPerTransaction = errorsAggResponse.aggregations.transactions.buckets.reduce(
+      (accumulatedErrorsPerTransaction, bucket: TraceErrorsAggBucket) => {
+        const {
+          key,
+          errors: { doc_count }
+        } = bucket;
+        return {
+          ...accumulatedErrorsPerTransaction,
+          [key]: doc_count
+        };
+      },
+      {}
+    );
```

Comment:
Personal nit: which of the two do find easier to read?

1) Current
```ts
const errorsPerTransaction = errorsAggResponse.aggregations.transactions.buckets.reduce(
  (accumulatedErrorsPerTransaction, bucket: TraceErrorsAggBucket) => {
    const {
      key,
      errors: { doc_count }
    } = bucket;
    return {
      ...accumulatedErrorsPerTransaction,
      [key]: doc_count
    };
  },
  {}
);

```

2) Adapted
```ts
const errorsPerTransaction = errorsAggResponse.aggregations.transactions.buckets.reduce(
  (acc, bucket: TraceErrorsAggBucket) => ({
      ...acc,
      [bucket.key]: bucket.errors.doc_count
  })
  {}
);
```

I realize this is very much personal pref.. I generally avoid the nested destructuring but I might be the odd one out :)

Reply:
I usually prefer nested destructuring followed by composition in smaller terms, but I don't feel strongly about it, so i can adapt. I'm happy to make these kinds of changes in the spirit of style cohesion and readability, but generally, I feel like code styles we want to enforce to should be codified in the linting step, otherwise it can be rather subjective.

Reply:
> but generally, I feel like code styles we want to enforce should be codified in the linting step, otherwise it can be rather subjective

Good point 👍 This made me curious to see if there was an ESLint rule for this. The closest I got was [this closed issue](https://github.com/eslint/eslint/issues/7973), so clearly there hasn't been overwhelming support for this idea.